src/ucp/core/ucp_worker.h

```diff
@@ -154,7 +154,7 @@ enum {
     ({ \
         ucp_mem_desc_t *rdesc = ucs_mpool_get_inline((_mp)); \
         if (rdesc != NULL) { \
-            VALGRIND_MAKE_MEM_DEFINED(rdesc, sizeof(*rdesc)); \
+            VALGRIND_MAKE_MEM_DEFINED(rdesc, (_mp)->data->elem_size - sizeof(ucs_mpool_elem_t)); \
```

Comment:
cuda pipelinine with tcp is complaining that stage buffer is pointed to uninitialized bytes.  this is because we are doing zcopy from CUDA memory to stage buffer and this is not recognized as a valid initialized buffer.

Reply:
how tcp is different? Why do we need stage buffer to be initialized?

Reply:
It seems Valgrind checks if TCP send buffer is initialized or not. In this case, buffer is initialized/staged from CUDA using cudaMemCpy().
maybe this is not a problem in IB case because of RDMA

```
==44069== Syscall param sendmsg(msg.msg_iov[2]) points to uninitialised byte(s)
==44069==    at 0x78DCC6D: ??? (in /usr/lib64/libpthread-2.17.so)
==44069==    by 0x4EA0A28: ucs_socket_do_iov_nb (sock.c:388)
==44069==    by 0x4EA0C33: ucs_socket_sendv_nb (sock.c:444)
==44069==    by 0x51D00C3: uct_tcp_ep_am_sendv (tcp_ep.c:1086)
==44069==    by 0x51D1870: uct_tcp_ep_put_zcopy (tcp_ep.c:1379)
==44069==    by 0x568985E: uct_ep_put_zcopy (uct.h:2333)
==44069==    by 0x568985E: ucp_rndv_progress_rma_put_zcopy_inner (rndv.c:880)
==44069==  Address 0x3d7040c0 is 8 bytes inside a block of size 524,296 client-defined
==44069==    at 0x4E84817: ucs_mpool_get_inline (mpool.inl:34)
==44069==    by 0x4E853D3: ucs_mpool_get (mpool.c:165)
==44069==    by 0x4E857E3: ucs_mpool_get_grow (mpool.c:242)
==44069==    by 0x56818AB: ucs_mpool_get_inline (mpool.inl:23)
==44069==    by 0x568E196: ucp_rndv_pipeline (rndv.c:1105)
==44069==    by 0x568F958: ucp_rndv_rtr_handler_inner (rndv.c:1243)
==44069==    by 0x568F958: ucp_rndv_rtr_handler (rndv.c:1202)
```

Reply:
looks like, it needs to be fixed in TCP transport. 
@dmitrygx, can you please check?

Reply:
@brminich I investigated whether this valgrind issue is related to TCP or not. I added the` `printf()` for the first byte of the UCP's buffer in `uct_ep_put_zcopy()` and Valgrind complains that this memory read was unitialized. The buffer that's passed to UCT PUT Zcopy comes from UCP memory pool (this is a staging buffer - UCP does `cudaMemcpy()` from GPU buffer to this HOST buffer and then does UCT PUT Zcopy for this HOST buffer). Looks like, Valgirnd doesn't understand that we initialized this HOST buffer by the data from GPU buffer when doing `cudaMemcpy()`.
I see that @bureddy's test compare target buffer and initiator buffer - and they are equal, so it means that Valgrind is wrong here and we should suppress this. I think that the @bureddy's fix is a good way to do it. Wdyt?

Reply:
but it is not really `defined` when we get it from mpool. Maybe set it `defined` in rndv pipelining, or better when CudaMemcpy is done?

Reply:
> but it is not really `defined` when we get it from mpool.

yes

> Maybe set it `defined` in rndv pipelining, or better when CudaMemcpy is done?

makes sense to me
wdyt about adding `VALGRIND_MAKE_MEM_DEFINED` for HOST buffer in `ucp_mem_type_pack`?

Reply:
i think we'd need to define it somewhere in cuda_copy transport, like we do it for gets in IB.
@bureddy, @Akshay-Venkatesh , i'd try to add `VALGRIND_MAKE_MEM_DEFINED` macro in `uct_cuda_copy_progress_event_queue` or in `uct_cuda_copy_ep_get_zcopy`. For the former we would need to update `uct_cuda_copy_event_desc_t` to include data length

Reply:
> @bureddy, @Akshay-Venkatesh , i'd try to add `VALGRIND_MAKE_MEM_DEFINED` macro in `uct_cuda_copy_progress_event_queue` or in `uct_cuda_copy_ep_get_zcopy`. For the former we would need to update `uct_cuda_copy_event_desc_t` to include data length

@dmitrygx 👍

Reply:
Fixed in uct_cuda_copy_ep_get_zcopy.

Reply:
Catching up a bit late. From what I understand, calling `VALGRIND_MAKE_MEM_DEFINED` on staging buffer about to passed to `cudaMemcpy` is equivalent to initializing it which avoids the issue of uninitialized bytes. Is that correct? If so, why do we do it per transfer attempt and not just do it for all the staging buffers at the time of cudaHostRegister or during memory pool init? Do we want to initialize per transfer?

Reply:
it is marking as defined for Valgrind.  I think it better here to just avoid suppressing real errors if making it in generic code